### ISSUE TITLE
Provide a way to use a bundled version of Chromium 51+ in medic-android

### ISSUE THREAD

**Issue by browndav** [medic/medic-webapp#2801]
Currently, medic-android relies on the system Webview. This means we get whatever version of Webview is provided by the system. On Android 5+, this component is automatically updated. On Android 4.4.x and lower, it is not. Currently the Tecno Y4 is Chromium 30 and does not get Webview updates.

Provide an option to use a bundled version of Chromium (or Chrome) inside of `medic-android`, in lieu of the system-provided Webview. This could be handled by linking to a prebuilt shared object that we then bundle with our APK, or it may involve building Chromium from source. Prefer the first method initially, if possible, just to make sure we don't invest a ton of time in building from source before we know the approach is viable.

Once this is complete, we'll have a couple new attacks on the storage quota problem. In any case, we'd be able to use Origin Trial keys for persistent storage (quota impact still TBD). If we build from source, we'd be able to directly patch the Chromium runtime to exempt us from the quota restrictions.

**Comment by alxndrsn** [comment]
Related: https://github.com/medic/medic-webapp/issues/997

**Comment by mandric** [comment]
There is a new branch over here of the chromium bundled app https://github.com/medic/medic-android/tree/crosswalk-2016 but @alxndrsn do you have any links to the deployed apks on s3?  Is there any reason we don't publish a link to them?

**Comment by mandric** [comment]
There is a chromium command line option labeled "for testing only" because it removes quotas for all origin domains, and this could be passed to Crosswalk, so we might not have to build chromium from source if we use this command line flag?
https://github.com/crosswalk-project/crosswalk-website/wiki/Use-Chromium-command-lines-in-your-apps-on-Android
http://peter.sh/experiments/chromium-command-line-switches/#unlimited-storage

**Comment by browndav** [comment]
I'm seeing the `--unlimited-storage` option work properly on Chromium `51.0.2704.63` (64-bit, Mac OS X). I'm able to hit a `QuotaExceededError`, restart with the option enabled, and see successful transaction commits once again.

**Comment by mandric** [comment]
Still no luck, I'm just modifying and testing the default project generated from `crosswalk-app`.  https://gist.github.com/mandric/7bf05e750d7ed3b5f907c4fb44abdd8e

**Comment by mandric** [comment]
Soliciting help over here:
https://lists.crosswalk-project.org/pipermail/crosswalk-help/2016-October/002274.html

**Comment by mandric** [comment]
Looks like we'll need to point crosswalk to a fork of chromium... https://github.com/crosswalk-project/crosswalk-website/wiki/Downstream-Chromium

**Comment by alxndrsn** [comment]
Building crosswalk is not documented for OSX; I recommend Ubuntu, although have not completed the process yet.

**Comment by alxndrsn** [comment]
@mandric have you tried passing chromium command line options via `src/main/assets/xwalk-command-line`, as documented at https://github.com/crosswalk-project/crosswalk-website/wiki/Use-Chromium-command-lines-in-your-apps-on-Android ?

**Comment by mandric** [comment]
Yes, tried all the methods. See ML post.

**Comment by alxndrsn** [comment]
# Ready for testing

* crosswalk-fork, managed from https://github.com/medic/crosswalk-build
* latest build of `medic-android`, including `XWalkWebview` from crosswalk version [`medic-0.4.0`](https://github.com/medic/crosswalk-build/tree/medic-0.4.0) is version [`0.1.175`](https://github.com/medic/medic-android/tree/v0.1.175) which implements the `--unlimited-storage` flag
* pull request for the `--unlimited-storage` flag to crosswalk at https://github.com/crosswalk-project/crosswalk/pull/3958

**Comment by sglangevin** [comment]
After doing some testing, it seems that updating to the new container will involve clearing all data and re-syncing. We have thousands of users in some of our deployments - is there anything we can do to allow users to update without having to re-download all of their data? cc @alxndrsn

**Comment by alxndrsn** [comment]
I can research this.  I doubt there's an API for getting access to the old WebView's data from the new WebView, but it might be possible to sync the databases by loading the two different WebViews and getting them to communicate through Java.  It would likely be a lot of work though, and from a dev perspective it would seem simpler to manage a staged rollout of the new version to the users, and take the hit on the extra data usage.
If we did implement automated migration, it would seem to make sense to use CouchDB's replication API, but this would require PouchDB in the XWalkWebView to function as replication server.  PouchDB may support this, but probably not in the standard build for web browsers.
@sglangevin it might be worth opening a separate issue to discuss this, unless we're definitely not going ahead.

**Comment by ngaruko** [comment]
Just wondering. Looking at the conversations above, did we come to a conclusion around this issue @sglangevin , @alxndrsn, @garethbowen  ?

**Comment by garethbowen** [comment]
@alxndrsn has been working on this but I'm not sure if there's a separate issue or not.

**Comment by sglangevin** [comment]
We don't have a separate issue AFAIK. There are some commits in a branch in medic-android related to the migration of data from one WebView to another. Perhaps we should create a new issue for that piece and close this one.

**Comment by sglangevin** [comment]
@ngaruko you should check in with @alxndrsn when he is back to see if we need to split the second part of this off into another issue. The original build Alex created was acceptance tested.

**Comment by alxndrsn** [comment]
Tracked at https://github.com/medic/medic-projects/issues/1246
